tests/ui/js/storyTellingServerUITests.js

```diff
@@ -69,98 +69,98 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                }]
+            },{
+                name: "Test themed page loading functions with server config values",
+                expect: 4,
+                sequence: [{
+                    task: "sjrk.storyTelling.storyTellingServerUiTester.loadClientConfigFromServer",
+                    args: ["/clientConfig", "{that}"],
                     resolve: "jqUnit.assertDeepEq",
-                    resolveArgs: ["The themed page load resolved as expected", loadThemedPageTestCases.learningReflections, "{arguments}.0"]
+                    resolveArgs: ["Custom theme was loaded successfully", "{that}.clientConfig", "{arguments}.0"]
                 },{
-                    funcName: "sjrk.storyTelling.storyTellingServerUiTester.assertCustomCssLoaded",
-                    args: ["learningReflections.css", 1]
+                    task: "sjrk.storyTelling.loadTheme",
+                    resolve: "jqUnit.assertDeepEq",
+                    resolveArgs: ["The themed page load resolved as expected", "{that}.clientConfig", "{arguments}.0"]
                 },{
-                    // test the CSS/JS injection function directly
-                    funcName: "sjrk.storyTelling.loadCustomThemeFiles",
-                    args: ["sjrk.storyTelling.testUtils.callbackVerificationFunction", {"theme": "learningReflections"}]
+                    funcName: "sjrk.storyTelling.storyTellingServerUiTester.verifyCustomCssLoaded",
+                    args: ["{that}.clientConfig.theme", 1]
                 },{
-                    funcName: "sjrk.storyTelling.storyTellingServerUiTester.assertCustomCssLoaded",
-                    args: ["learningReflections.css", 2]
+                    task: "sjrk.storyTelling.loadCustomThemeFiles",
+                    args: ["{that}.clientConfig"],
+                    resolve: "sjrk.storyTelling.storyTellingServerUiTester.verifyCustomCssLoaded",
+                    resolveArgs: ["{that}.clientConfig.theme", 2]
                 }]
             }]
         }]
     });
 
-    sjrk.storyTelling.storyTellingServerUiTester.loadThemedPageSingleTest = function (callback, expectedTheme) {
-        var loadPromise = fluid.promise();
-
-        sjrk.storyTelling.storyTellingServerUiTester.setupMockServer("/clientConfig", JSON.stringify({
-            theme: expectedTheme,
-            baseTheme: "base",
-            authoringEnabled: true
-        }));
-
-        sjrk.storyTelling.loadThemedPage(callback).then(function (clientConfig) {
-            loadPromise.resolve(clientConfig);
-        }, function () {
-            loadPromise.reject();
+    sjrk.storyTelling.storyTellingServerUiTester.loadClientConfigFromServer = function (url, testerComponent) {
+        var configPromise = fluid.promise();
+
+        $.get(url).then(function (data) {
+            if (data.theme !== data.baseTheme) {
+                testerComponent.clientConfig.theme = data.theme;
+                configPromise.resolve(data);
```

Comment:
It looks like you are passing in the `sjrk.storyTelling.storyTellingServerUiTester` component and assigning the `theme` property of its `clientConfig` member with the `theme` property from the returned `data` object. But in the test, you do a `jqUnit.assertDeepEq` with `clientConfig` member against the `data` object returned by the resolved promise. I'm not sure of the validity of this test, or perhaps what you are trying to test may not be clear. It would seem that if the `theme` is invalid or possibly even missing that the test would still pass. If these are tests against a live instance, which I believe they are, we are making an assumption that the configuration for `baseTheme` and `authoringEnabled` hasn't been changed. 

I see in the tests that you follow up this one with a call to `sjrk.storyTelling.loadTheme` which also fetches the config from the server. So I'm assuming what you are trying to do is to pre-fetch the config data to use as the expected value to compare the returned data from `sjrk.storyTelling.loadTheme` against. If this is correct, here are some thoughts:
- They are essentially using the same mechanism for fetching this data, `$.get("/clientConfig")`. Can you source the data by another means; for example, can you fetch and source the data directly out of `sjrk.storyTelling.server.config.json5` instead?
- If you are just pre-fetching why not assign the whole `data` object to the `clientConfig` member?
  - If you can do this, then you can skip the assertion and just make the resolver a call to `fluid.set` and assign the value there instead of in the `sjrk.storyTelling.storyTellingServerUiTester.loadClientConfigFromServer` function.

Considerations for future JIRAs, if they don't already exist:
- `sjrk.storyTelling.loadTheme` shouldn't hardcode the URL it uses to fetch the clientConfig
- It might be a good idea not to serve the tests because we'll need to be careful not to write any tests that might modify the server. Also we probably don't want to serve the tests on a production instance.

Reply:
Fetching the `clientConfig` data directly from the configuration file is not possible, as these are client-side tests that aim to ensure these server-interfacing functions are working as expected. Exposing the config file would not be any different, functionally speaking, from the `/clientConfig` endpoint already present in the server.

With that in mind, I agree that checking that a value I've loaded from the server against another value loaded from exactly the same part of the server is invalid; at best, it's redundant. I will remove this part and implement the bypass you've suggested.

Regarding serving the tests outside of the development environment, there are currently no tests that make any changes to the server and no tests that affect the database directly. Can you elaborate on why we wouldn't want to expose the tests on the production container? Is it to avoid effects on that environment? FWIW the same methods/endpoints that can be harnessed by the tests are also usable/callable by any user at this point.

Reply:
What I was suggesting about the alternate path for fetching the config file is that you ensure that the file is being retrieved correctly. That is, you verify the correct file is being fetched via that endpoint. I agree that it would be better as a server specific test.

If I'm remembering correctly, another issue, which is kind of separate, is that you can't run the server tests when an instance of the server is running, because they use the same port. This prevents running all the tests at once. What would be nice is if you could run the server, integration, and browser tests independently, but also run them all together using our testem setup.

Regarding the unit tests in production. Partly it's something that you probably don't want to expose to a user. It could be a security risk if it is leaking access or information about secured access to the server. I was more worried about accidentally writing a test that modified the live server, especially with integration tests. It sounds like the tests currently aren't doing this, but it's something that we'll have to be aware of.